myia_backend_python/myia_backend_python/python.py

```diff
@@ -0,0 +1,838 @@
+        for node in toposort(graph.output, NodeVisitor(), in_graph(graph)):
+            if self._add_node(node):
+                seq.append(node)
+        output = []
+        # Get code for each node.
+        for op in seq:
+            op_name = self.local_ref(op)
+            if op.is_apply():
+                op_code = self.on_apply(op)
+            elif op.is_constant_graph():
+                op_code = self.on_function(op.value, op)
+            elif op.is_constant():
+                op_code = self.on_constant(op)
+            else:
+                raise AssertionError(f"Unsupported node: {op}")
+            # If latest op is graph.output, we can write code
+            # to return it immediately.
+            if op is seq[-1] is graph.output:
+                prefix = "return"
+            else:
+                prefix = f"{op_name} ="
+            # Code for a node may be a list of lines of code.
+            # In such case, latest code line should be only the return value.
+            if isinstance(op_code, list):
+                build_code = op_code[:-1]
+                return_code = op_code[-1]
+                output.extend(build_code)
+                output.append(f"{prefix} {return_code}")
+            elif op_code:
+                output.append(f"{prefix} {op_code}")
+        # Output may be empty, for e.g. if function just returns a parameter.
+        if not output:
+            output.append(f"return {self.ref(graph.output)}")
+
+        constants_code = []
+        closures_code = []
+        for cst_name, cst_val in self.const_name_to_value.items():
+            if not self._is_inline_const(cst_val):
+                # Write only non-inlinable constants.
+                # Other are inlined, hence it's useless to define them.
+                constants_code.append(f"{cst_name} = {cst_val}")
+        if constants_code:
+            constants_code = ["# Constants"] + constants_code + [""]
+        for fn_name, (fn_params, fn_body) in self.closure_name_to_code.items():
+            fn_signature = f"def {fn_name}({', '.join(fn_params)}):"
+            closures_code.append(fn_signature)
+            closures_code.append(fn_body)
+        # Body code contains constants, then closures, then function code.
+        return param_names, constants_code + closures_code + output
+
+
+class PythonCompiler(_Compiler):
+    """Compile a myia graph into a Python function."""
+
+    REG_PYTHON_NAME = re.compile(r"^[A-Za-z_][A-Za-z0-9_]*$")
+    REG_INVALID_CHARS = re.compile(r"[^A-Za-z0-9_]+")
+    REG_INVALID_START = re.compile(r"^[^A-Za-z0-9_]+")
+    REG_INVALID_END = re.compile(r"[^A-Za-z0-9_]+$")
+    REG_DIGIT_START = re.compile(r"^[0-9]")
+
+    def __init__(self):
+        """Initialize."""
+        self._prim_graph_cache = {}
+        self.make_const = PythonConstantConverter()
+        self.graph_to_name = {}
+        self.fn_name_to_code = {}
+        self.node_labeler = NodeLabeler(relation_symbols=short_relation_symbols)
```

Comment:
```suggestion
        self.node_labeler = NodeLabeler()
```

Why are you passing `short_relation_symbols` here? Without a `relation_symbols` table, it'll generate stuff like `2:grad_fprop:if_true:sumtree`, so you just need to replace the `:`. Otherwise, you're losing information by generating Unicode symbols and then just dropping them.

Reply:
Hi ! Indeed, I can remove short_relation_symbols, but it will then generate very long variable names, e.g. something like `_copy_copy_copy_grad_bprop_app_copy_copy_copy_opt_copy__apply0`. Is it right ?

Reply:
Oh yeah good point. You can pass `{"copy": "", "opt": ""}` instead of `short_relation_symbols` to avoid displaying the `copy` and `opt` transforms. Basically pass only the entries of `short_relation_symbols` that map to `""`. Then the above would become simply `_grad_bprop_app_apply0`.

Reply:
If you can just make the above adjustment to avoid displaying the copy/opt transforms in the names, I think it's all good for me.

Reply:
Done !